File: measuretool/MeasureTools/CalculateOverlap.m
function [ration, overlapdot] = CalculateOverlap(elp1,elp2)
% 计算两个椭圆的重叠面积
% elp格式为[epl(1),elp(2)]为椭圆中心点，[elp1(3),elp(4)]为椭圆半长短轴,elp(5)为弧度制旋转角

%% 形状参数转换为椭圆的一般方程
parm1 = ELPShape2Equation(elp1);
parm2 = ELPShape2Equation(elp2);

%% 计算椭圆Y方向的搜索范围
y1_min = elp1(2) - max(elp1(3:4)); y1_max = elp1(2) + max(elp1(3:4));
y2_min = elp2(2) - max(elp2(3:4)); y2_max = elp2(2) + max(elp2(3:4));
y_min = floor(max([y1_min,y2_min]));
y_max = ceil(min([y1_max,y2_max]));
search_step = 0.2;

%% 计算重叠面积
S12 = 0; % S12表示重叠面积
overrange = zeros(length(y_min:search_step:y_max),3); idx_lap = 1;


for i = y_min:search_step:y_max
    [x11,x12] = CalculateRangeAtY(parm1,i);
    [x21,x22] = CalculateRangeAtY(parm2,i);
%     x11
%     x12
%     x21
%     x22
%     i

    %disp(num2str([x11,x12,x21,x22]));
    rangetemp = [-1,-2,i];
    if ~isempty(x11) && ~isempty(x21)
        if x11<=x21 && x12>=x21 
            if x12<x22   % [x11   [x21   x12]  x22]
                S12 = S12 + x12-x21;
                rangetemp =[x12,x21,i];
            else   % [x11   [x21     x22]  x12]
                S12 = S12 + x22-x21;
                rangetemp =[x22,x21,i];
            end
        elseif x21<=x11 && x22>=x11  
            if x22 < x12   % [x21   [x11   x22]   x12]  
                S12 = S12 + x22 - x11;
                rangetemp =[x22,x11,i];
            else    % [x21   [x11     x12]   x22] 
                S12 = S12 + x12-x11;
                rangetemp =[x12,x11,i];
            end
        end
    end
    overrange(idx_lap,:) = rangetemp;
    idx_lap = idx_lap + 1;
end

ration = S12 * search_step/(pi*elp1(3)*elp1(4)+pi*elp2(3)*elp2(4) - S12 * search_step);
overlapdot = overrange;
end

File: measuretool/MeasureTools/CalculateOverlap.cpp
#include "MeasureTools.h"
#include "mex.h"

#include <iostream>
#include <vector>
using namespace std;

#define PI 3.14159265358979

void CalculateOverlap(double *elp1, double *elp2, double *ration, vector<double> *overlapdot)
{
	double parm1[6], parm2[6];
	ELPShape2Equation(elp1, parm1);
	ELPShape2Equation(elp2, parm2);
	
	double y1_min, y1_max, y2_min, y2_max, y_min, y_max;
	y1_min = elp1[1] - fmax(elp1[2], elp1[3]); y1_max = elp1[1] + fmax(elp1[2], elp1[3]);
	y2_min = elp2[1] - fmax(elp2[2], elp2[3]); y2_max = elp2[1] + fmax(elp2[2], elp2[3]);
	y_min = floor(fmax(y1_min, y2_min));
	y_max = ceil(fmin(y1_max, y2_max));

	double search_step = 0.2;
	int idx_lap = 1;
	double S12 = 0;

	for (double i = y_min; i <= y_max+1e-6; i = i + search_step)
	{
		double x11, x12, x21, x22;
		CalculateRangeAtY(parm1, i, &x11, &x12);
		CalculateRangeAtY(parm2, i, &x21, &x22);

		//mexPrintf("[%.4f,%.4f],[%.4f,%.4f]\n", x11, x12, x21, x22);

		double rangetemp[3] = { -1,-2,i };
		if (x11 <= x12&& x21 <= x22)
		{
			if (x11 <= x21 && x12 >= x21)
			{
				if (x12 < x22)
				{
					S12 += x12 - x21;
					rangetemp[0] = x12, rangetemp[1] = x21;
				}
				else
				{
					S12 += x22 - x21;
					rangetemp[0] = x22, rangetemp[1] = x21;
				}
			}
			else if (x21 <= x11 && x22 >= x11)
			{
				if (x22 < x12)
				{
					S12 += x22 - x11;
					rangetemp[0] = x22, rangetemp[1] = x11;
				}
				else
				{
					S12 += x12 - x11;
					rangetemp[0] = x12, rangetemp[1] = x11;
				}
			}
		}

		overlapdot[0].push_back(rangetemp[0]);
		overlapdot[1].push_back(rangetemp[1]);
		overlapdot[2].push_back(rangetemp[2]);
	}

	//mexPrintf("%.4f\n", S12);
	*ration = S12 *search_step / (PI*elp1[2] * elp1[3] + PI*elp2[2] * elp2[3] - S12*search_step);

	

}